File: dailyQuestion/2020-06/06-09/python/solution_dp.py

# 从左向右遍历
def translateNum(num: int) -> int:
    s = str(num)
    f_1 = f_2 = 1
    for i in range(2, len(s) + 1):
        pre = s[i - 2:i]
        f = f_1 + f_2 if pre >= "10" and pre <= "25" else f_1
        f_2 = f_1
        f_1 = f
    return f_1

# 从右向左遍历
def translateNum_2(num: int) -> int:
    s = str(num)
    f_1 = f_2 = 1
    for i in range(2, len(s) + 1):
        pre = s[i - 2:i]
        f = f_1 + f_2 if pre >= "10" and pre <= "25" else f_1
        f_2 = f_1
        f_1 = f
    return f_1

if __name__ == '__main__' :
    num = 12258
    result = translateNum(num)
    print(result)

File: dailyQuestion/2020/2020-06/06-09/golang/solution_dp.go
package main

import (
	"fmt"
	"strconv"
)

func main() {
	num := 12258
	result := translateNum(num)
	fmt.Println(result)

}

func translateNum(num int) int {
	f_2, f_1 := 1, 1
	s := strconv.Itoa(num)
	for i := 2; i <= len(s); i++ {
		pre := s[i-2 : i]
		f := 0
		if pre >= "10" && pre <= "25" {
			f = f_1 + f_2
		} else {
			f = f_1
		}
		f_2 = f_1
		f_1 = f
	}
	return f_1
}
